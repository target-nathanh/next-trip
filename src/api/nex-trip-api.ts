import { Direction, NexTripResult, Place, Route, Vehicle } from '../types';

const API_ROOT = 'https://svc.metrotransit.org/nextripv2';
async function getRoutes(): Promise<Route[]> {
  return baseFetch<Route[]>(`${API_ROOT}/routes`);
}

async function getDirectionsForRoute(routeId: string): Promise<Direction[]> {
  return baseFetch<Direction[]>(`${API_ROOT}/directions/${routeId}`);
}

async function getPlaces(routeId: string, directionId: number): Promise<Place[]> {
  return baseFetch<Place[]>(`${API_ROOT}/stops/${routeId}/${directionId}`);
}

async function getNexTripResult(
  routeId: string,
  directionId: number,
  placeCode: string
): Promise<NexTripResult> {
  return baseFetch<NexTripResult>(`${API_ROOT}/${routeId}/${directionId}/${placeCode}`);
}

async function getNexTripResultByStopId(stopId: string): Promise<NexTripResult> {
  return baseFetch<NexTripResult>(`${API_ROOT}/${stopId}`);
}

async function getVehiclesByRoute(routeId: string): Promise<Vehicle[]> {
  return baseFetch<Vehicle[]>(`${API_ROOT}/vehicles/${routeId}`);
}

async function getDirectionsAndPlaces(
  routeId: string,
  directionId: number
): Promise<{ places: Place[]; directions: Direction[] }> {
  const directions = await getDirectionsForRoute(routeId);
  const places = await getPlaces(routeId, directionId);

  return { places, directions };
}

async function baseFetch<T>(url: string, body?: any, extraHeaders = {}): Promise<T> {
  const fetchConfig: { body?: string; headers: Record<string, string> } = {
    headers: {
      'Content-Type': 'application/json',
      ...extraHeaders,
    },
  };
  if (body) {
    fetchConfig.body = JSON.stringify(body);
  }
  const response = await fetch(url, fetchConfig);

  if (response.ok) {
    return await response.json();
  }

  const error = new Error(response.status.toString());
  error.message = response.statusText;
  throw error;
}

export const NexTripApi = {
  getRoutes,
  getDirectionsForRoute,
  getNexTripResult,
  getPlaces,
  getNexTripResultByStopId,
  getVehiclesByRoute,
  getDirectionsAndPlaces,
};
