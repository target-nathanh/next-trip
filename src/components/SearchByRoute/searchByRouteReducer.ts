import { Direction, Place, Route } from '../../types';

interface SearchByRouteReducerState {
  directions?: Direction[];
  routes?: Route[];
  places?: Place[];
  selectedRouteId?: string;
  selectedDirectionId?: number;
  selectedPlaceId?: string;
}

export function searchByRouteReducer(
  state: SearchByRouteReducerState,
  action: { type: string; payload?: any }
): SearchByRouteReducerState {
  switch (action.type) {
    case 'SET_PLACES_AND_DIRECTIONS': {
      const payload = action.payload;
      return {
        ...state,
        ...payload,
      };
    }
    case 'SET_ROUTES':
      return {
        ...state,
        routes: action.payload.routes,
      };
    case 'SET_PLACES':
      return {
        ...state,
        places: action.payload.places,
      };
    case 'SET_DIRECTIONS':
      return {
        ...state,
        directions: action.payload.directions,
      };
    case 'SET_SELECTED_ROUTE':
      return {
        ...state,
        selectedRouteId: action.payload.selectedRouteId,
      };
    case 'SET_SELECTED_DIRECTION':
      return {
        ...state,
        selectedDirectionId: action.payload.selectedDirectionId,
      };
    case 'SET_SELECTED_PLACE':
      return {
        ...state,
        selectedPlaceId: action.payload.selectedPlaceId,
      };
    case 'CLEAR_STATE':
      return {
        ...state,
        directions: [],
        places: [],
        selectedPlaceId: '',
        selectedDirectionId: -1,
      };
    default:
      throw new Error();
  }
}
