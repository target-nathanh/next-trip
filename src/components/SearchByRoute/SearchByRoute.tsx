import React, { useEffect, useReducer } from 'react';
import RoutesDropdown from '../RoutesDropdown/RoutesDropdown';
import { RouteParams } from '../../types';
import { NexTripApi } from '../../api/nex-trip-api';
import DirectionsDropdown from '../DirectionsDropdown/DirectionsDropdown';
import PlacesDropdown from '../PlacesDropdown/PlacesDropdown';
import { Grid } from '@material-ui/core';
import Results from '../Results/Results';
import {
  Route as RouterRoute,
  Switch,
  useHistory,
  useRouteMatch,
  matchPath,
} from 'react-router-dom';
import { searchByRouteReducer } from './searchByRouteReducer';

const SearchByRoute: React.FC = () => {
  const { path, url } = useRouteMatch();
  const history = useHistory();
  const match = matchPath<RouteParams>(history.location.pathname, {
    path: '/route/:routeId/:directionId/:placeCode',
  });
  const [state, dispatch] = useReducer(searchByRouteReducer, {
    directions: [],
    places: [],
    routes: [],
    selectedPlaceId: '',
    //Defaulting to -1 because some directionIds are 0, so in order to test for them in boolean statements I want to check that the value is greater than -1
    selectedDirectionId: -1,
    selectedRouteId: '',
  });

  useEffect(() => {
    async function getRoutes() {
      if (match && match.params.routeId && match.params.directionId && match.params.placeCode) {
        const result = await NexTripApi.getDirectionsAndPlaces(
          match.params.routeId,
          parseInt(match.params.directionId, 10)
        );
        dispatch({
          type: 'SET_PLACES_AND_DIRECTIONS',
          payload: {
            places: result.places,
            directions: result.directions,
            selectedRouteId: match.params.routeId,
            selectedDirectionId: parseInt(match.params.directionId, 10),
            selectedPlaceId: match.params.placeCode,
          },
        });
      } else {
        dispatch({ type: 'CLEAR_STATE' });
      }
      const routes = await NexTripApi.getRoutes();
      dispatch({ type: 'SET_ROUTES', payload: { routes } });
    }

    getRoutes();
  }, []);

  useEffect(() => {
    async function getDirections() {
      if (state.selectedRouteId) {
        const retrievedDirections = await NexTripApi.getDirectionsForRoute(state.selectedRouteId);
        dispatch({ type: 'SET_DIRECTIONS', payload: { directions: retrievedDirections } });
      }
    }
    getDirections();
  }, [state.selectedRouteId]);

  useEffect(() => {
    async function getPlaces() {
      if (state.selectedRouteId && state.selectedDirectionId && state.selectedDirectionId > -1) {
        const retrievedPlaces = await NexTripApi.getPlaces(
          state.selectedRouteId,
          state.selectedDirectionId
        );
        dispatch({ type: 'SET_PLACES', payload: { places: retrievedPlaces } });
      }
    }
    getPlaces();
  }, [state.selectedRouteId, state.selectedDirectionId]);

  const onPlaceSelected = (placeId: string) => {
    dispatch({ type: 'SET_SELECTED_PLACE', payload: { selectedPlaceId: placeId } });
    const getResults = async () => {
      if (
        state.selectedRouteId &&
        state.selectedDirectionId &&
        state.selectedDirectionId > -1 &&
        placeId
      ) {
        const nexTripResult = await NexTripApi.getNexTripResult(
          state.selectedRouteId,
          state.selectedDirectionId,
          placeId
        );
        history.push({
          pathname: `${url}/${state.selectedRouteId}/${state.selectedDirectionId}/${placeId}`,
          state: { results: nexTripResult },
        });
      }
    };
    getResults();
  };

  return (
    <Grid container direction="column" justify="center" alignItems="center">
      <RoutesDropdown
        routes={state.routes}
        selectedRouteId={state.selectedRouteId}
        onSelectRoute={(routeId: string) => {
          dispatch({ type: 'SET_SELECTED_ROUTE', payload: { selectedRouteId: routeId } });
          dispatch({ type: 'CLEAR_STATE' });
        }}
      />
      {state.selectedRouteId && (
        <DirectionsDropdown
          directions={state.directions}
          selectedDirectionId={state.selectedDirectionId}
          onSelectDirection={(directionId: number) =>
            dispatch({
              type: 'SET_SELECTED_DIRECTION',
              payload: { selectedDirectionId: directionId },
            })
          }
        />
      )}
      {state.selectedDirectionId && state.selectedDirectionId > -1 && state.selectedRouteId && (
        <PlacesDropdown
          places={state.places}
          selectedPlaceId={state.selectedPlaceId}
          onSelectPlace={(placeId: string) => onPlaceSelected(placeId)}
        />
      )}

      <Switch>
        <RouterRoute path={`${path}/:routeId/:directionId/:placeCode`}>
          <Results />
        </RouterRoute>
      </Switch>
    </Grid>
  );
};

export default SearchByRoute;
