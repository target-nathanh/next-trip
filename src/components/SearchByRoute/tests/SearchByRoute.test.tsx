import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { NexTripApi } from '../../../api/nexTripApi';
import SearchByRoute from '../SearchByRoute';
import { MemoryRouter } from 'react-router-dom';
import mockRoutes from '../../../mockData/routesMock.json';
import mockPlaces from '../../../mockData/placesMock.json';
import mockDirections from '../../../mockData/directionsMock.json';

describe('SearchByRoute', () => {
  test('matches snapshot for default state', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/route']}>
        <SearchByRoute />
      </MemoryRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('calls the api with the selected stop id on clicking the icon', async () => {
    NexTripApi.getRoutes = jest.fn().mockResolvedValue(mockRoutes);
    NexTripApi.getPlaces = jest.fn().mockResolvedValue(mockPlaces);
    NexTripApi.getDirectionsForRoute = jest.fn().mockResolvedValue(mockDirections);
    NexTripApi.getNexTripResult = jest.fn();
    const { getByRole, getAllByRole } = render(
      <MemoryRouter initialEntries={['/route']}>
        <SearchByRoute />
      </MemoryRouter>
    );
    fireEvent.mouseDown(getByRole('button'));

    const selectedRoute = mockRoutes[1];
    const selectedDirection = mockDirections[1];
    const selectedPlace = mockPlaces[1];

    const routeOption = await screen.findByText(selectedRoute.route_label);
    fireEvent.click(routeOption);
    expect(NexTripApi.getDirectionsForRoute).toHaveBeenCalledWith(selectedRoute.route_id);

    // Directions
    let buttons = getAllByRole('button');
    expect(buttons.length).toEqual(2);
    fireEvent.mouseDown(buttons[1]);

    const directionOption = await screen.findByText(selectedDirection.direction_name);
    fireEvent.click(directionOption);
    expect(NexTripApi.getPlaces).toHaveBeenCalledWith(
      selectedRoute.route_id,
      selectedDirection.direction_id
    );

    // Places
    buttons = getAllByRole('button');
    expect(buttons.length).toEqual(3);
    fireEvent.mouseDown(buttons[2]);

    const placesOption = await screen.findByText(selectedPlace.description);
    fireEvent.click(placesOption);
    expect(NexTripApi.getNexTripResult).toHaveBeenCalledWith(
      selectedRoute.route_id,
      selectedDirection.direction_id,
      selectedPlace.place_code
    );
  });
});
