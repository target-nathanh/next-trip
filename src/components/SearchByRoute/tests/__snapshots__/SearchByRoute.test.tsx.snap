// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchByRoute matches snapshot for default state 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
>
  <div
    class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-select-1"
  >
    <div
      aria-haspopup="listbox"
      class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      role="button"
      tabindex="0"
    >
      <span>
        â€‹
      </span>
    </div>
    <input
      aria-hidden="true"
      class="MuiSelect-nativeInput"
      placeholder="Select route"
      tabindex="-1"
      value=""
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
  
</div>
`;
